Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs

```diff
@@ -0,0 +1,463 @@
+        }
+    }
+
+    public void OnHeadRevAdmin(Mind.Mind mind, IPlayerSession headRev)
+    {
+        var revRule = EntityQuery<RevolutionaryRuleComponent>().FirstOrDefault();
+        if (revRule == null)
+        {
+            GameTicker.StartGameRule("Revolutionary");
+        }
+        if (mind.OwnedEntity != null)
+        {
+            EnsureComp<RevolutionaryRuleComponent>(mind.OwnedEntity.Value);
+            GiveHeadRevRole(mind, headRev);
+            AssignCommandStaff();
+        }
+    }
+    private void OnMobStateChanged(EntityUid uid, RevolutionaryRuleComponent comp, MobStateChangedEvent ev)
+    {
+        if (ev.NewMobState == MobState.Dead || ev.NewMobState == MobState.Invalid)
+            CheckFinish();
+    }
+    /// <summary>
+    /// Checks if all Head Revs are dead and if all command is dead to either end the round or remove all revs. Or both.
+    /// </summary>
+    private void CheckFinish()
+    {
+        var stunTime = TimeSpan.FromSeconds(4);
+        var headJobs = _prototypeManager.Index<DepartmentPrototype>("Command");
+        var secJobs = _prototypeManager.Index<DepartmentPrototype>("Security");
+
+        var query = AllEntityQuery<RevolutionaryRuleComponent, GameRuleComponent>();
+        while (query.MoveNext(out var uid, out var revs, out var gameRule))
+        {
+            if (!GameTicker.IsGameRuleAdded(uid, gameRule))
+                continue;
+            var headRevs = EntityQuery<HeadRevolutionaryComponent, MobStateComponent>(true);
+            var inRound = 0;
+            var dead = 0;
+            foreach (var rev in headRevs)
+            {
+                if (rev.Item2.CurrentState == MobState.Dead || rev.Item2.CurrentState == MobState.Invalid)
+                {
+                    dead++;
+                }
+                inRound++;
+            }
+
+            // If no Head Revs are alive all normal Revs will lose their Rev status and rejoin Nanotrasen
+            if (dead == inRound)
+            {
+                revs.RevsLost = true;
+                var rev = AllEntityQuery<RevolutionaryComponent>();
+                while (rev.MoveNext(out var id, out var comp))
+                {
+                    if (HasComp<RevolutionaryComponent>(id) && !HasComp<HeadRevolutionaryComponent>(id))
+                    {
+                        var name = Identity.Entity(id, EntityManager);
+                        _npcFaction.AddFaction(id, "NanoTrasen");
+                        _npcFaction.RemoveFaction(id, "Revolutionary");
+                        _sharedStun.TryParalyze(id, stunTime, true);
+                        RemCompDeferred<RevolutionaryComponent>(id);
+                        _popup.PopupEntity(Loc.GetString("rev-break-control", ("name", name)), id);
+                        _adminLogManager.Add(LogType.Mind, LogImpact.Medium, $"{ToPrettyString(id)} was deconverted due to all Head Revolutionaries dying.");
+                    }
+                }
+            }
```

Comment:
Move this to a generic method so it's useable.

Reply:
I understand making the other stuff you said into methods but I'm not sure why I would make this one a useable method or how as I feel this is pretty specific to revs.

Reply:
Putting everything in one god function for "CheckFinish" means you can't do anything cleanly as this is just trying to remove revolutionary.